lib/Conversion/CombToCGGI/CombToCGGI.cpp

```diff
@@ -151,6 +180,55 @@ struct ConvertTruthTableOp : public OpConversionPattern<TruthTableOp> {
   }
 };
 
+// ConvertSecretCastOp removes secret.cast operations between multi-bit secret
+// integers and tensors of single-bit secrets.
+struct ConvertSecretCastOp : public OpConversionPattern<secret::CastOp> {
+  ConvertSecretCastOp(mlir::MLIRContext *context)
+      : OpConversionPattern<secret::CastOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      secret::CastOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    // If this is a cast from secret<i8> to secret<tensor<8xi1>> or vice
+    // versa, replace with the cast's input.
+    auto inputTy =
+        cast<secret::SecretType>(op.getInput().getType()).getValueType();
+    auto outputTy =
+        cast<secret::SecretType>(op.getOutput().getType()).getValueType();
+
+    if (auto inputIntTy = dyn_cast<IntegerType>(inputTy)) {
+      if (auto outputTensorTy = dyn_cast<RankedTensorType>(outputTy)) {
+        auto outputTensorElt =
+            dyn_cast<IntegerType>(outputTensorTy.getElementType());
+        assert(outputTensorElt &&
+               "expected underlying integer type of secret tensor");
+        if (inputIntTy.getWidth() ==
+            outputTensorTy.getNumElements() * outputTensorElt.getWidth()) {
+          rewriter.replaceOp(op, op.getInput());
```

Comment:
Shouldn't this be the adaptor's input? You want the value after it's been type-converted, right?

Reply:
Yeah - It still works out (I wonder if the pattern applicator ends up converting it again...)

Reply:
Updated